### ISSUE TITLE
Some question about the query and target alignment

### ISSUE THREAD

**Issue by cocoshe** [lucas-ventura/CoVR#38]
I try to train and reproduce the training process with blip2 coco pretrained model.
And I try to calculate recall on train/val/test dataset.(Yes, including the training part)
 
- after epoch 1:
```
# train set
{'R1': 5.27, 'R5': 16.56, 'R10': 22.38, 'R50': 39.48, 'R_mean': 20.92}
# val set
{'R1': 75.7812, 'R5': 96.875, 'R10': 99.2188, 'R50': 100.0, 'R_mean': 90.625}
# test set
{'R1': 56.37, 'R5': 81.72, 'R10': 89.22, 'R50': 97.97, 'meanR3': 75.77, 'meanR4': 81.32, 'annotation': 'webvid8m-covr_test.csv'}
```

- after epoch 2:
```
# train set
{'R1': 6.69, 'R5': 19.31, 'R10': 25.83, 'R50': 44.47, 'R_mean': 24.07}
# val set
{'R1': 77.3438, 'R5': 98.4375, 'R10': 99.2188, 'R50': 100.0, 'R_mean': 91.6667}
# test set
{'R1': 58.98, 'R5': 83.36, 'R10': 90.08, 'R50': 98.24, 'meanR3': 77.47, 'meanR4': 82.67, 'annotation': 'webvid8m-covr_test.csv'}
```
......
- after epoch 5:
```
# train set
{'R1': 7.37, 'R5': 20.84, 'R10': 27.71, 'R50': 47.14, 'R_mean': 25.77}
# val set
{'R1': 79.6875, 'R5': 98.4375, 'R10': 99.2188, 'R50': 100.0, 'R_mean': 92.4479}
# test set
{'R1': 59.61, 'R5': 84.14, 'R10': 91.09, 'R50': 98.36, 'meanR3': 78.28, 'meanR4': 83.3, 'annotation': 'webvid8m-covr_test.csv'}
```

The final loss is about `Loss: 9.279157`

So I'm curious why the recall on training set is so low, bacause we train on the training set, it should get the highest recall scores. This is quite strange.

**Comment by lucas-ventura** [comment]
Can you please provide the exact training command and parameters you are using?
In particular, what batch size are you running with? The performance is quite sensitive to this setting.

**Comment by cocoshe** [comment]
> Can you please provide the exact training command and parameters you are using? In particular, what batch size are you running with? The performance is quite sensitive to this setting.

Here's my training command: `python train.py model=blip2-coco model/ckpt=blip2-l-coco test=webvid-covr`
I train it with blip2, on covr dataset. I didn't change any hyperparams. 

So what I've done?

1. In the train loop
I modify the model, return the `features`, to record the query emb and target emb in the training set when training.
```python
    calculate_recall = True
    
    # Collect features for recall calculation if needed
    if calculate_recall:
        all_query_feats = []
        all_tar_img_feats = []
        all_tar_txt_feats = []
    
    for batch_idx, batch in enumerate(train_loader):
        optimizer.zero_grad()
        
        if calculate_recall:
            loss, features = model(batch, fabric, return_features=True)   ###### Here, I return the query emb and target emb of this batch
            
            # Collect features for recall calculation (detach to avoid memory issues)
            all_query_feats.append(features['query_si_feat'].detach().cpu())
            all_tar_img_feats.append(features['tar_img_feat'].detach().cpu())
```

My `BLIP2Cir` class forward is here:
```
    def forward(self, batch, fabric, return_features=False):
        ref_img = batch["ref_img"]
        tar_img_feat = batch["tar_img_feat"]
        caption = batch["edit"]

        ref_img.half()

        device = ref_img.device

        # Encode the target image
        tar_img_feat = tar_img_feat.to(device)
        tar_img_feat = concat_all_gather(tar_img_feat, fabric)

        # Text
        text_tokens = self.tokenizer(
            caption,
            padding="max_length",
            truncation=True,
            max_length=self.max_txt_len,
            return_tensors="pt",
        ).to(device)

        if self.train_vit:
            ref_img_embs = self.ln_vision(self.visual_encoder(ref_img))
        else:
            with torch.no_grad():
                ref_img_embs = self.ln_vision(self.visual_encoder(ref_img))

        # Encode the reference image
        ref_img_atts = torch.ones(ref_img_embs.size()[:-1], dtype=torch.long).to(device)

        ###============== Image-text Matching ===================###
        query_tokens = self.query_tokens.expand(ref_img_embs.shape[0], -1, -1)
        query_atts = torch.ones(query_tokens.size()[:-1], dtype=torch.long).to(
            self.device
        )
        attention_mask = torch.cat([query_atts, text_tokens.attention_mask], dim=1)

        output = self.Qformer.bert(
            text_tokens.input_ids,  # [bs, 32]
            query_embeds=query_tokens,  # [bs, 32, 768]
            attention_mask=attention_mask,  # [bs, 64]
            encoder_hidden_states=ref_img_embs,  # [bs, 677, 1408]
            encoder_attention_mask=ref_img_atts,  # [bs, 677]
            return_dict=True,
        )

        vl_embs = output.last_hidden_state[:, : query_tokens.size(1), :]
        query_si_feat = F.normalize(self.text_proj(vl_embs), dim=-1)
        query_si_feat = all_gather_with_grad(query_si_feat, fabric)

        # mean over all query tokens
        query_si_feat = query_si_feat.mean(dim=1)
        tar_img_feat = tar_img_feat.mean(dim=1)

        # s=source, t=target, i=image, c=caption, w=weight
        loss = 0
        
        if self.si_ti_weight > 0:
            si_ti_loss = self.loss(query_si_feat, tar_img_feat, self.temp)
            loss += si_ti_loss * self.si_ti_weight

        if self.si_tc_weight > 0:
            assert "tar_txt_feat" in batch, "tar_txt_feat is not in batch"
            tar_txt_feat = batch["tar_txt_feat"]

            tar_txt_feat = all_gather_with_grad(tar_txt_feat, fabric)

            si_tc_loss = self.loss(query_si_feat, tar_txt_feat, self.temp)
            loss += si_tc_loss * self.si_tc_weight
##################################### Here's my modification #############################
        if return_features:
            # Return both loss and features for recall calculation
            features = {
                'query_si_feat': query_si_feat,
                'tar_img_feat': tar_img_feat,
            }
            
            if self.si_tc_weight > 0:
                features['tar_txt_feat'] = tar_txt_feat
                
            return loss, features   #### I return the embs of training set
        else:
            # Return only loss for backward compatibility
            return loss
###################################################################################    
```



2. I move the test process into the epoch for loop, so I can get the test recall every epoch.(In the repo now, the test process only occurs after all the 5 epochs finish)
```python
    for epoch in range(cfg.trainer.max_epochs):
        scheduler(optimizer, epoch)

        columns = shutil.get_terminal_size().columns
        fabric.print("-" * columns)
        fabric.print(f"Epoch {epoch + 1}/{cfg.trainer.max_epochs}".center(columns))

        train(model, loader_train, optimizer, fabric, epoch, cfg)

        if cfg.val:
            fabric.print("Evaluate")
            instantiate(cfg.evaluate, model, loader_val, fabric=fabric)

        # Test on all test datasets after each epoch
        fabric.print("Testing")
        for dataset in cfg.test:
            columns = shutil.get_terminal_size().columns
            fabric.print("-" * columns)
            fabric.print(f"Testing on {cfg.test[dataset].dataname} (Epoch {epoch + 1})".center(columns))

            data = instantiate(cfg.test[dataset])
            test_loader = fabric.setup_dataloaders(data.test_dataloader())

            test = instantiate(cfg.test[dataset].test)
            test(model, test_loader, fabric=fabric)

        state = {
            "epoch": epoch,
            "model": model,
            "optimizer": optimizer,
            "scheduler": scheduler,
        }
        if cfg.trainer.save_ckpt == "all":
            fabric.save(f"ckpt_{epoch}.ckpt", state)
        elif cfg.trainer.save_ckpt == "last":
            fabric.save("ckpt_last.ckpt", state)

        fabric.barrier()
```

And as for val set, it occurs every epoch, so easily get the recall.(The repo now already support this, it will eval every epoch and print the recall)


That's how I get the recalls on the three datasets.

Intuitively, we train on the train set, and in the default setting I found it using only the target video as the supervision, we can see the config here:

https://github.com/lucas-ventura/CoVR/blob/f335ebc3ddf1cb26106a3aaf374cf15901c59986/configs/train.yaml#L27

So the training and valuation and testing should be aligned, which means we take `ref img` and `edit` as input, `target video` as gt. But don't know why the training recall is so low, maybe I misunderstand something, do you have any idea, or what I did wrong when calculating the recall on train set?

**Comment by cocoshe** [comment]
On the three sets, there are just little margin on test set(`R1` about 59, close to the result in the paper), and the val recall is better(`R1` about 79) so I think the training process should be correct to reproduce the paper. But the train set recall scores are abnormal, don't know why...

**Comment by lucas-ventura** [comment]
I didn’t look at your code in detail, but the most probable explanation is that on the training set there are many more target embeddings compared to val/test. This means you have a lot more distractors, and often embeddings that are very similar to each other, which makes retrieval harder and lowers the recall scores. There are about 1.6M triplets in the training set vs ~2.5k in the test set, so it’s normal that training recall looks much lower — the retrieval task is simply much more difficult.

I’m sure that if you visualize the results (e.g., check a few queries where it “fails”), you’ll see that the retrieved examples still make sense qualitatively, even if the strict recall metric is low.

**Comment by cocoshe** [comment]
> I didn’t look at your code in detail, but the most probable explanation is that on the training set there are many more target embeddings compared to val/test. This means you have a lot more distractors, and often embeddings that are very similar to each other, which makes retrieval harder and lowers the recall scores. There are about 1.6M triplets in the training set vs ~2.5k in the test set, so it’s normal that training recall looks much lower — the retrieval task is simply much more difficult.
> 
> I’m sure that if you visualize the results (e.g., check a few queries where it “fails”), you’ll see that the retrieved examples still make sense qualitatively, even if the strict recall metric is low.

OK I will try it later. And do you remember the training loss? `Loss: 9.279157` in my reproducing, it seems kind of large, is it normal?

**Comment by lucas-ventura** [comment]
Unfortuntly, I don't remember the training loss value.

**Comment by cocoshe** [comment]
> I didn’t look at your code in detail, but the most probable explanation is that on the training set there are many more target embeddings compared to val/test. This means you have a lot more distractors, and often embeddings that are very similar to each other, which makes retrieval harder and lowers the recall scores. There are about 1.6M triplets in the training set vs ~2.5k in the test set, so it’s normal that training recall looks much lower — the retrieval task is simply much more difficult.
> 
> I’m sure that if you visualize the results (e.g., check a few queries where it “fails”), you’ll see that the retrieved examples still make sense qualitatively, even if the strict recall metric is low.


> but the most probable explanation is that on the training set there are many more target embeddings compared to val/test. This means you have a lot more distractors, and often embeddings that are very similar to each other, which makes retrieval harder and lowers the recall scores.

That's true!

I try to just use a slice of training set, like `[:256]`, and get the accuracy upto `0.90+`. I also try some CLIP-like model to calculate the similarity matrix and get the accuracy on some other datasets. (different model, different datasets), and even the slice increase to 2000 samples `[:2000]`, the another CLIP-like model on its dataset still get `0.99+` accuracy, but CoVR, this model, on the CoVR train set, will decrease to around`0.60`. So maybe there are more similar samples in training set of this CoVR dataset, or maybe there are still some gap between model stability, which may indicates there are more progress could be made for the society development in the future.

However, really thanks for your sincerely and timely help! 
REALLY GREAT WORK AND COMMUNICATION
